composer/algorithms/seq_length_warmup/seq_length_warmup.py

```diff
@@ -217,33 +211,78 @@ def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
                     'Sequence Length Warmup requires the `state.dataloader` to have a `batch_size` attribute.') from e
             if per_gpu_macrobatch is None:
                 raise RuntimeError('Sequence Length Warmup algorithm requires constant batch size.')
-            per_gpu_batch = ceil(per_gpu_macrobatch / state.grad_accum)
-
-            input_ids = torch.randint(low=0,
-                                      high=vocab_size - 1,
-                                      size=(per_gpu_batch, self.max_seq_length),
-                                      device=device).long()
-            labels = input_ids.clone()
-            attn_mask = torch.ones_like(labels)
-            model_inputs = {
-                'input_ids': input_ids,
-                'labels': labels,
-                'attention_mask': attn_mask,
-            }
-
-            # start by running a forward and backward pass
-            # of the maximum sequence length to allocate cache.
-            with get_precision_context(state.precision):
-                outputs = state.model.forward(model_inputs)
-                loss = self._original_model.loss(outputs, model_inputs)
-
-            # since use_grad_scaling is in the Trainer, and we
-            # don't care about the loss values, skip scaling
-            for loss_item in ensure_tuple(loss):
-                loss_item.backward()
-
-            for optimizer in state.optimizers:
-                optimizer.zero_grad()
+
+            batch_clone = {}
+            device_batch_size = 0
+            device = None
+            for k, v in state.batch.items():
+                # Truncate any sequence-shaped tensors to at most the max_seq_length sequence.
+                # Assume the second dimension is always sequence position.
+                if len(v.shape) > 1:
+                    seq_dim = min(self.max_seq_length, v.shape[1])
+                    batch_clone[k] = torch.clone(v[:, :seq_dim])
+                else:
+                    batch_clone[k] = torch.clone(v)
+                device_batch_size = v.shape[0]
+                device = v.device
+
+            grad_accum_successful = False
+            while not grad_accum_successful:
```

Comment:
The code looks ok, but i'm not the expert. It's sufficiently long that might want to break out into its own helper function for modularity. 

I'm also curious how well this works. We've heard sporadic reports of auto grad accum not freeing up all the memory. Would be interesting to experiment with/without using `torch.cuda.empty_cache()`

Reply:
I played around a bit with `torch.cuda.empty_cache()` and couldn't find any clear advantages. It is possible (likely, even) that any memory issues with the auto grad accum will also occur here, since the logic is essentially identical.

At the very least, I took your suggestion wrt using a helper function.